README.md
# connector-azure-active-directory


release_notes.md
# What's new in 2.1.0:

- List Groups: Retrieves the list of groups from Azure Active Directory with filtering capability
- Get Group Details: Retrieves the details of a particular group.
- List SignIns Events: Retrieves the list sign ins events from Azure Active Directory based on the input parameters such as the query for filtering users you have specified.
- Add Member: Add member to a group.
- Remove Member: Remove member from a group.

azure-active-directory/connector.py
""" Copyright start
  Copyright (C) 2008 - 2021 Fortinet Inc.
  All rights reserved.
  FORTINET CONFIDENTIAL & FORTINET PROPRIETARY SOURCE CODE
  Copyright end """

from connectors.core.connector import Connector, get_logger, ConnectorError
from .operations import operations, _check_health

logger = get_logger('azure-active-directory')


class AzureAcivedirectory(Connector):
    def execute(self, config, operation, params, **kwargs):
        try:
            connector_info = {"connector_name": self._info_json.get('name'),
                              "connector_version": self._info_json.get('version')}
            operation = operations.get(operation)
        except Exception as err:
            logger.exception(err)
            raise ConnectorError(err)
        return operation(config, params, connector_info)

    def check_health(self, config):
        logger.info('starting health check')
        connector_info = {"connector_name": self._info_json.get('name'),
                          "connector_version": self._info_json.get('version')}
        _check_health(config, connector_info)
        logger.info('completed health check no errors')


azure-active-directory/operations.py
""" Copyright start
  Copyright (C) 2008 - 2021 Fortinet Inc.
  All rights reserved.
  FORTINET CONFIDENTIAL & FORTINET PROPRIETARY SOURCE CODE
  Copyright end """

from requests import request, exceptions as req_exceptions
from .microsoft_api_auth import *
from urllib import parse
from requests_toolbelt.utils import dump
import copy

logger = get_logger('azure-active-directory')

API_VERSION = "v1.0"


def api_request(method, endpoint, connector_info, config, params=None, data=None, headers={}):
    try:
        ms = MicrosoftAuth(config)
        endpoint = ms.host + "/" + API_VERSION + endpoint
        token = ms.validate_token(config, connector_info)

        headers['Authorization'] = token
        headers['Content-Type'] = 'application/json'
        headers['consistencylevel'] = 'eventual'
        try:
            response = request(method, endpoint, headers=headers, params=params, json=data, verify=ms.verify_ssl)
            logger.debug('\n{}\n'.format(dump.dump_all(response).decode('utf-8')))
            if response.status_code in [200, 201, 204]:
                if response.text != "":
                    return response.json()
                else:
                    return True
            else:
                if response.text != "":
                    err_resp = response.json()
                    failure_msg = err_resp['error']['message']
                    error_msg = 'Response [{0}:{1} Details: {2}]'.format(response.status_code, response.reason,
                                                                         failure_msg if failure_msg else '')
                else:
                    error_msg = 'Response [{0}:{1}]'.format(response.status_code, response.reason)
                logger.error(error_msg)
                raise ConnectorError(error_msg)
        except req_exceptions.SSLError:
            logger.error('An SSL error occurred')
            raise ConnectorError('An SSL error occurred')
        except req_exceptions.ConnectionError:
            logger.error('A connection error occurred')
            raise ConnectorError('A connection error occurred')
        except req_exceptions.Timeout:
            logger.error('The request timed out')
            raise ConnectorError('The request timed out')
        except req_exceptions.RequestException:
            logger.error('There was an error while handling the request')
            raise ConnectorError('There was an error while handling the request')
        except Exception as err:
            raise ConnectorError(str(err))
    except Exception as err:
        raise ConnectorError(str(err))


def _fetch_remaining_pages(response, url_params, connector_info, config):
    url_params.update({"$top":1000})
    results = copy.deepcopy(response)
    while '@odata.nextLink' in response:
        skiptoken = parse.parse_qs(parse.urlparse(response['@odata.nextLink']).query)['$skiptoken'][0]
        url_params.update({"$skiptoken": skiptoken})
        response = api_request("GET", "/auditLogs/signIns", connector_info, config, params=url_params)
        results['value'] += response['value']
        logger.debug('Append {} more records'.format(str(len(response['value']))))
    return results


def list_users(config, params, connector_info):
    try:
        search = params.get('$search')
        if search:
            search = '\"{0}\"'.format(search)
        payload = {
            '$filter': params.get('$filter'),
            '$search': search,
            '$select': params.get('$select')
        }
        payload = {k: v for k, v in payload.items() if v is not None and v != ''}
        response = api_request("GET", "/users", connector_info, config, params=payload)
        return response
    except Exception as err:
        raise ConnectorError(str(err))


def _list_records(config, params, connector_info, endpoint):
    try:
        url_params = {"$filter": params.get('$filter')} if params.get('$filter', None) else {}
        get_all_pages = params.get("get_all_pages")
        if params.get("$top"):
            url_params.update({"$top": params.get('$top')})
        if params.get("$skipToken"):
            url_params.update({"$skipToken": params.get('$skipToken')})
        response = api_request("GET", endpoint, connector_info, config, params=url_params)
        if '@odata.nextLink' in response and get_all_pages:
            return _fetch_remaining_pages(response, url_params, connector_info, config)
        else:
            return response

    except Exception as err:
        raise ConnectorError(str(err))


def list_groups(config, params, connector_info):
    return _list_records(config, params, connector_info, "/groups")


def list_sign_ins(config, params, connector_info):
    return _list_records(config, params, connector_info, "/auditLogs/signIns")


def get_group_details(config, params, connector_info):
    try:
        response = api_request("GET", "/groups/{0}".format(params.get('id')), connector_info, config)
        return response
    except Exception as err:
        raise ConnectorError(str(err))


def remove_member(config, params, connector_info):
    try:
        response = api_request("DELETE", "/groups/{0}/members/{1}/$ref".format(params.get('id'),
                                                                               params.get("dir_object_id")),
                                                                               connector_info, config)
        return response
    except Exception as err:
        raise ConnectorError(str(err))

def add_member(config, params, connector_info):
    try:
        payload = {
            "@odata.id": "https://graph.microsoft.com/v1.0/directoryObjects/{0}".format(params.get("dir_object_id"))
        }
        response = api_request("POST", "/groups/{0}/members/$ref".format(params.get('id')), connector_info, config,
                               data=payload)
        return response
    except Exception as err:
        raise ConnectorError(str(err))

def get_user_details(config, params, connector_info):
    try:
        response = api_request("GET", "/users/{0}".format(params.get('id')), connector_info, config)
        return response
    except Exception as err:
        raise ConnectorError(str(err))


def enable_user(config, params, connector_info):
    try:
        payload = {
            'accountEnabled': True
        }
        response = api_request("PATCH", "/users/{0}".format(params.get('id')), connector_info, config, data=payload)
        if response:
            return {'status': 'success', 'result': 'User account enable successfully'}
    except Exception as err:
        raise ConnectorError(str(err))


def disable_user(config, params, connector_info):
    try:
        payload = {
            'accountEnabled': False
        }
        response = api_request("PATCH", "/users/{0}".format(params.get('id')), connector_info, config, data=payload)
        if response:
            return {'status': 'success', 'result': 'User account disable successfully'}
    except Exception as err:
        raise ConnectorError(str(err))


def reset_password(config, params, connector_info):
    try:
        force_change = params.pop('force_change', False)
        payload = {'passwordProfile': {'forceChangePasswordNextSignIn': force_change}}
        password = params.get('password')
        if password:
            payload['passwordProfile']['password'] = password
        response = api_request("PATCH", "/users/{0}".format(params.get('id')), connector_info, config, data=payload)
        if response:
            return {'status': 'success', 'result': 'Password successfully reset'}
    except Exception as err:
        raise ConnectorError(str(err))


def add_user(config, params, connector_info):
    try:
        force_change = params.pop('force_change', False)
        password = params.pop('password')
        payload = {'passwordProfile': {'forceChangePasswordNextSignIn': force_change, 'password': password}}
        params.update(payload)
        additional_fields = params.pop('additional_fields', {})
        if additional_fields:
            params.update(additional_fields)
        response = api_request("POST", "/users", connector_info, config, data=params)
        if response:
            return response
    except Exception as err:
        raise ConnectorError(str(err))


def delete_user(config, params, connector_info):
    try:
        response = api_request("DELETE", "/users/{0}".format(params.get('id')), connector_info, config)
        if response:
            return {'status': 'success', 'result': 'User successfully Deleted'}
    except Exception as err:
        raise ConnectorError(str(err))


def _check_health(config, connector_info):
    try:
        return check(config, connector_info)
    except Exception as err:
        raise ConnectorError(str(err))


operations = {
    'list_users': list_users,
    'get_user_details': get_user_details,
    'enable_user': enable_user,
    'disable_user': disable_user,
    'reset_password': reset_password,
    'add_user': add_user,
    'delete_user': delete_user,
    'list_sign_ins': list_sign_ins,
    'list_groups': list_groups,
    'get_group_details': get_group_details,
    'remove_member':remove_member,
    'add_member':add_member
}


azure-active-directory/microsoft_api_auth.py
""" Copyright start
  Copyright (C) 2008 - 2021 Fortinet Inc.
  All rights reserved.
  FORTINET CONFIDENTIAL & FORTINET PROPRIETARY SOURCE CODE
  Copyright end """

from requests import request
from time import time, ctime
from os import path
from adal import AuthenticationContext
from datetime import datetime
from configparser import RawConfigParser
from base64 import b64encode, b64decode
from connectors.core.connector import get_logger, ConnectorError

logger = get_logger('azure-active-directory')

CONFIG_SUPPORTS_TOKEN = True
try:
    from connectors.core.utils import update_connnector_config
except:
    CONFIG_SUPPORTS_TOKEN = False
    configfile = path.join(path.dirname(path.abspath(__file__)), 'config.conf')

REFRESH_TOKEN_FLAG = False

# authorization types
AUTH_BEHALF_OF_USER = "On behalf of User - Delegate Permission"
AUTH_USING_APP = "Without a User - Application Permission"

# redirect url
DEFAULT_REDIRECT_URL = 'https://localhost/myapp'

# grant types
CLIENT_CREDENTIALS = 'client_credentials'
AUTHORIZATION_CODE = 'authorization_code'
REFRESH_TOKEN = 'refresh_token'


class MicrosoftAuth:

    def __init__(self, config):
        self.client_id = config.get("client_id")
        self.client_secret = config.get("client_secret")
        self.verify_ssl = config.get('verify_ssl')
        self.host = config.get("resource")
        if self.host[:7] == "http://":
            self.host = "https://{0}".format(self.host)
        elif self.host[:8] == "https://":
            self.host = "{0}".format(self.host)
        else:
            self.host = "https://{0}".format(self.host)
        tenant_id = config.get('tenant_id')
        self.auth_url = 'https://login.microsoftonline.com/{0}'.format(tenant_id)
        # depend on the auth type we add the class members.
        self.auth_type = config.get("auth_type")
        if self.auth_type == AUTH_BEHALF_OF_USER:
            self.refresh_token = ""
            self.code = config.get("code")
            self.scope = 'https://graph.microsoft.com/.default'
            self.token_url = "https://login.microsoftonline.com/{0}/oauth2/v2.0/token".format(tenant_id)
            if not config.get("redirect_url"):
                self.redirect_url = DEFAULT_REDIRECT_URL
            else:
                self.redirect_url = config.get("redirect_url")

    def convert_ts_epoch(self, ts):
        datetime_object = datetime.strptime(ctime(ts), "%a %b %d %H:%M:%S %Y")
        return datetime_object.timestamp()

    def encode_token(self, token):
        try:
            token = token.encode('UTF-8')
            return b64encode(token)
        except Exception as err:
            logger.error(err)

    def generate_token(self, REFRESH_TOKEN_FLAG):
        try:
            if not self.auth_type == AUTH_BEHALF_OF_USER:
                ac = AuthenticationContext(self.auth_url)
                token_resp = ac.acquire_token_with_client_credentials(self.host, self.client_id, self.client_secret)
                ts = token_resp.get('expiresOn')
                datetime_object = datetime.strptime(ts, '%Y-%m-%d %H:%M:%S.%f')
                token_resp['expiresOn'] = datetime_object.timestamp()
                return token_resp
            else:
                resp = self.acquire_token_on_behalf_of_user(REFRESH_TOKEN_FLAG)
                ts_now = time()
                resp['expiresOn'] = (ts_now + resp['expires_in']) if resp.get("expires_in") else None
                resp['accessToken'] = resp.get("access_token")
                resp.pop("access_token")
                return resp

        except Exception as err:
            try:
                logger.error("{0}".format(err.error_response['error_description']))
                raise ConnectorError("{0}".format(err.error_response['error_description']))
            except:
                logger.error("{0}".format(err))
                raise ConnectorError("{0}".format(err))

    def write_config(self, token_resp, config, section_header):
        time_key = ['expiresOn']
        token_key = ['accessToken']

        config.add_section(section_header)
        for key, val in token_resp.items():
            if key not in time_key and key not in token_key:
                config.set(section_header, str(key), str(val))
        for key in time_key:
            config.set(section_header, str(key), self.convert_ts_epoch(token_resp['expiresOn']))
        for key in token_key:
            config.set(section_header, str(key), self.encode_token(token_resp[key]).decode('utf-8'))

        try:
            with open(configfile, 'w') as fobj:
                config.write(fobj)
                fobj.close()
            return config
        except Exception as err:
            logger.error("{0}".format(str(err)))
            raise ConnectorError("{0}".format(str(err)))

    def handle_config(self, section_header, flag=False):
        # Lets setup the config parser.
        config = RawConfigParser()
        try:
            if path.exists(configfile) is False:
                token_resp = self.generate_token(REFRESH_TOKEN_FLAG)
                return self.write_config(token_resp, config, section_header)
            else:
                # Read existing config
                config.read(configfile)
                # Check for user
                if not config.has_section(section_header) and not flag:
                    # Write new config
                    token_resp = self.generate_token(REFRESH_TOKEN_FLAG)
                    return self.write_config(token_resp, config, section_header)
                else:
                    if flag:
                        config.remove_section(section_header)
                        with open(configfile, "w") as f:
                            config.write(f)
                    else:
                        config.read(config)
                return config

        except Exception as err:
            logger.error("Handle_config:Failure {0}".format(str(err)))
            raise ConnectorError(str(err))

    def validate_token(self, connector_config, connector_info):
        if CONFIG_SUPPORTS_TOKEN:
            ts_now = time()
            if not connector_config.get('accessToken'):
                logger.error('Error occurred while connecting server: Unauthorized')
                raise ConnectorError('Error occurred while connecting server: Unauthorized')
            expires = connector_config['expiresOn']
            expires_ts = self.convert_ts_epoch(expires)
            if ts_now > float(expires_ts):
                REFRESH_TOKEN_FLAG = True
                logger.info("Token expired at {0}".format(expires))
                self.refresh_token = connector_config["refresh_token"]
                token_resp = self.generate_token(REFRESH_TOKEN_FLAG)
                connector_config['accessToken'] = token_resp['accessToken']
                connector_config['expiresOn'] = token_resp['expiresOn']
                connector_config['refresh_token'] = token_resp.get('refresh_token')
                update_connnector_config(connector_info['connector_name'], connector_info['connector_version'],
                                         connector_config,
                                         connector_config['config_id'])

                return "Bearer {0}".format(connector_config.get('accessToken'))
            else:
                logger.info("Token is valid till {0}".format(expires))
                return "Bearer {0}".format(connector_config.get('accessToken'))
        else:
            client_id = connector_config.get('client_id')
            section_header = 'Microsoft-API-Auth-{0}'.format(client_id)
            time_key = ['expiresOn']
            token_key = ['accessToken']
            try:
                config = self.handle_config(section_header)
                ts_now = time()
                expires = config.get(section_header, 'expiresOn')
                if ts_now > float(expires):
                    REFRESH_TOKEN_FLAG = True
                    self.refresh_token = config.get(section_header, 'refresh_token')
                    logger.info("Token expired at {0}".format(str(expires)))
                    new_token = self.generate_token(REFRESH_TOKEN_FLAG)
                    for key, val in new_token.items():
                        if key in time_key:
                            config.set(section_header, str(key), self.convert_ts_epoch(new_token.get(key)))
                        if key in token_key:
                            config.set(section_header, str(key), self.encode_token(new_token[key]).decode('utf-8'))

                    with open(configfile, 'w') as fobj:
                        config.write(fobj)
                else:
                    logger.info("Token is valid till {0}".format(str(expires)))

                encoded_token = config.get(section_header, 'accessToken')
                decoded_token = b64decode(encoded_token.encode('utf-8'))
                token = "Bearer {0}".format(decoded_token.decode('utf-8'))
                return token
            except Exception as err:
                logger.error("{0}".format(str(err)))
                raise ConnectorError("{0}".format(str(err)))

    def remove_config(self):
        try:
            section_header = 'Microsoft-API-Auth-{0}'.format(self.client_id)
            self.handle_config(section_header, flag=True)
        except Exception as err:
            logger.error("{0}".format(str(err)))
            raise ConnectorError("{0}".format(str(err)))

    def acquire_token_on_behalf_of_user(self, REFRESH_TOKEN_FLAG):
        try:
            post_data = {
                "client_id": self.client_id,
                "scope": self.scope,
                "client_secret": self.client_secret,
                "redirect_uri": self.redirect_url
            }

            if not REFRESH_TOKEN_FLAG:
                post_data["grant_type"] = AUTHORIZATION_CODE,
                post_data["code"] = self.code
            else:
                post_data['grant_type'] = REFRESH_TOKEN,
                post_data['refresh_token'] = self.refresh_token

            response = request("POST", self.token_url, data=post_data, verify=self.verify_ssl)
            if response.status_code in [200, 204, 201]:
                return response.json()

            else:
                if response.text != "":
                    error_msg = ''
                    err_resp = response.json()
                    if err_resp and 'error' in err_resp:
                        failure_msg = err_resp.get('error_description')
                        error_msg = 'Response {0}: {1} \n Error Message: {2}'.format(response.status_code,
                                                                                     response.reason,
                                                                                     failure_msg if failure_msg else '')
                    else:
                        err_resp = response.text
                else:
                    error_msg = '{0}:{1}'.format(response.status_code, response.reason)
                raise ConnectorError(error_msg)

        except Exception as err:
            logger.error("{0}".format(str(err)))
            raise ConnectorError(error_msg)


def check(config, connector_info):
    try:
        ms = MicrosoftAuth(config)
        if CONFIG_SUPPORTS_TOKEN:
            if not 'accessToken' in config:
                token_resp = ms.generate_token(REFRESH_TOKEN_FLAG)
                config['accessToken'] = token_resp.get('accessToken')
                config['expiresOn'] = token_resp.get('expiresOn')
                config['refresh_token'] = token_resp.get('refresh_token')
                update_connnector_config(connector_info['connector_name'], connector_info['connector_version'], config,
                                         config['config_id'])
                return True
            else:
                token_resp = ms.validate_token(config, connector_info)
                return True
        else:
            ms.remove_config()
            client_id = config.get('client_id')
            section_header = 'Microsoft-API-Auth-{0}'.format(client_id)
            ms.handle_config(section_header)
            return True
    except Exception as err:
        raise ConnectorError(str(err))
